Suggestion: user-defined functions

pontusmelke opened the pull request:
User defined functions are a simpler form of procedures which always return a single value. This is possible to do with normal procedures by returning a single-item `Stream`. However doing so adds a lot of boiler plate both in how you call the procedure, how you create the procedure. Furthermore the planner gets no help in figuring out the cardinality of the query when using procedures.

A user-defined function is defined by 

``` java
package com.example;
....
@UserFunction
public String addFoo(@Name("input") String input)
{
    return input + "foo";
}
```

Functions can take any number of arguments, but always return a single value. Supported input and output types are, `long, Long, double, Double, Number, boolean, Boolean, String, Node, Relationship, Path`, and `List<T>` where `T` is one of the supported types (or a list of one of the supported types, or a list of a list of one of the ... you get it!) or `Map<String, Object>`.

From Cyphers point-of-view this is just a normal function, so you can call it with

``` cypher
MATCH (n) RETURN com.example.addFoo(n.name)
```

Basically anything you can do with a normal function you can do with a user-defined function.

User functions are read-only i.e. you can inject a `GraphDataBaseService` but you cannot perform any writes nor schema operations.

Random decisions made while implementing that are up for discussion:
- This might be an awful idea?
- There is a separate `@UserFunction` annotation, but it is more or less a copy of `@Procedure` alternatively `@Procedure` could be used for functions as well and the compiler would need to be smarter and look at the signature to figure out if the intention is to create a function or a procedure. This mostly felt wrong to me, not sure if we want them always to be identical. However having two annotations adds some code duplication.
- Not allowed to create functions in the root namespace, mostly to avoid collisions with built-in functions but if we think it is important to create such functions we can check for collisions in other ways.
- ~~Allowing functions to have side-effects such as creating nodes, relationships and doing schema operations might also be an awful idea?~~
- ~~`@Function` is very generic and collides with `java.util.function.Function`, however this not a problem if you use arrow syntax.~~

pontusmelke committed:
Added namespace to function calls

pontusmelke committed:
Added mechanism for function lookup in planning

pontusmelke committed:
Added repository and basic functions in kernel

pontusmelke committed:
Added capability to execute udf in cypher

pontusmelke committed:
Make function be callable from Cypher

- Introduce the necessary SPIs for calling functions
- Introduce command for executing functions in Cypher

pontusmelke committed:
Added annotation and parser for functions

pontusmelke committed:
Dissallow functions in root namespace

chrisvest commented:
I wouldn't allow functions to have dependencies injected. That way side-effects are limited and don't have to be accounted for in the planner.

jexp commented:
@chrisvest but then you couldn't do even simple things.
Like getting a node by it's id or getting labels from the database as a function.

Otherwise you can't really limit sideeffects as Node/Relationship have getGraphDatabaseService();

And you can do anything on the JVM anyway and you can easily equip a function class with a static variable (or access one somewhere else that provides you access to the Kernel or whatever).

So while it would annoy users it wouldn't really prevent anyone who wanted to, to access whatever API they want.

jexp commented:
- This might be an awful idea?
  -> I think it's a great idea and much needed. Thanks for spending time building it.
- There is a separate @Function annotation, but it is more or less a copy of @Procedure alternatively @Procedure could be used for functions as well and the compiler would need to be smarter and look at the signature to figure out if the intention is to create a function or a procedure. This mostly felt wrong to me, not sure if we want them always to be identical. 

However having two annotations adds some code duplication.

-> I think a separate annotation is much better
- Not allowed to create functions in the root namespace, mostly to avoid collisions with built-in functions but if we think it is important to create such functions we can check for collisions in other ways.

-> that's a good idea and forces people to scope their functions
- Allowing functions to have side-effects such as creating nodes, relationships and doing schema operations might also be an awful idea?

-> good question, functions with side-effects are tricky and probably not needed, we can start without side effects and read-only functions and gather feedback.

@Function is very generic and collides with java.util.function.Function, however this not a problem if you use arrow syntax.

-> we could use @UserDefinedFunction

boggle commented:
Hey great suggestion and a great addition to the existing procedure work! :)

A few remarks:
- I think we should overhaul our approach to annotations and pick a common prefix. `@User{Procedure|Function}`. Other options might be `@Cypher{Procedure|Function}` even `@Custom{Procedure|Function}`.
- Perhaps we should also consider using an annotation for declaring default values again instead of duplicating that functionality in each annotation (Procedure, Function, ...) we add.
- I think it is paramount to enforce that functions are immutable and do not perform any updates. Anything else breaks Cypher's idea of expression evaluation. Updates are much better performed by (possibly VOID) procedures. The fact that user code can break free from these shackles is similar how they could also grab dependency injector and run amok with that. When you deploy custom code we don't stop you from shooting your own feet - but we don't suggest you do that either. Let's please not break the semantic model of Cypher and make it clear that functions are not supposed to mutate the graph and enforce that by only injecting a GDB with limited access. We already have the required facilities for that in the procedure code so doing that shouldn't be a big deal.

pontusmelke commented:
Build failures are unrelated as far as I know.
- I think having functions being read-only makes sense, I'll change so that functions only can read from the database.
- I don't mind `UserFunction` but we can't change `Procedure` -> `UserProcedure` without going through a full deprecation cycle, right @boggle ? Unfortunately supporting both `UserProcedure` and `Procedure` is annoying since annotations can't really share code. However if we want to do it we should just do it but then the deprecation should probably go into 3.0.

boggle commented:
@pontusmelke I think we could easily add a second (new) annotation and deprecate the existing one. Could duplication shouldn't be a big issue as annotations are very tiny java classes and the only other additional work would be for ReflectiveProcedureCompiler to pick out both.

boggle commented:
@pontusmelke We're also in the process of finishing up the procedures CIP after that fact. We should probably write one for functions as well.

pontusmelke commented:
@boggle Ok, will rename to `UserFunction`. Renaming and deprecating procedures will be done in separate work. Functions doesn't add anything to Cypher, not sure if we need a CIP.

To consider is also that procedures are not only _user procedures_, we also have the built-in procedures that are not really _user procedures_.

pontusmelke committed:
Changed functions to be read-only

pontusmelke committed:
Changed Function -> UserFunction

pontusmelke commented:
I have fixed most comments:
- Functions are now read-only
- `Function` -> `UserFunction`
- However I chose not support defaults on functions since we're not supporting them for normal Cypher functions. The default values lives together with `@Name` annotation which is only used for Procedures.

jexp commented:
Thanks for addressing them so quickly. Good stuff. 

Defaults would still be _very useful_ as we don't have overloading 

Von meinem iPhone gesendet

> Am 05.09.2016 um 12:05 schrieb Pontus Melke notifications@github.com:
> 
> I have fixed most comments:
> 
> Functions are now read-onlu
> Function -> UserFunction
> However I chose not support defaults on functions since we're not supporting them for normal Cypher functions. The default values lives together with @Name annotation which is only used for Procedures.
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub, or mute the thread.

boggle commented:
@pontusmelke why did you choose to not use @Name for functions? I don't see a reason to have that be different from procedures.

pontusmelke commented:
@boggle the `@Name` was added for procedures in order to have named input fields for procedures so that you could bind parameters from the context, such as with `CALL my.proc, {"procArg": 42}`. I don't see the need for having named input parameters in functions until we support things like `MATCH (n) RETURN id(node = n)`, as it is now it is just forcing users to add an extra annotation.

But if you feel strongly about it I can be persuaded to add it, anyone else having opinions on `@Name`?

jexp commented:
@Name would be nice for error reporting but not required.

Defaults OTOH would be really helpful.

pontusmelke committed:
Added @Name to functions and reverted name change

pontusmelke commented:
@jexp @boggle I am somewhat convinced, I added `@Name` (with the defaults and all) for functions as well. I still feel we force a lot of boiler plate on implementers, but maybe we can make `@Name` be optional on both procedures and functions in a separate piece of work?

jexp commented:
Yes I think in general we can infer parameter names via [reflection since Java8](http://docs.oracle.com/javase/8/docs/api/java/lang/reflect/Executable.html#getParameters--)

I'm for making it optional.

pontusmelke commented:
@jexp no the parameter names are only available if you compile with debug flags, that was the initial reason why `@Name` was added, however if the user doesn't use `@Name` we could maybe just call them `in1, in2, ...` or something. However this will not be part of this PR, for now it is mandatory on both functions and procedures.

boggle commented:
Awesome progress.. regarding using Java 8 reflection to infer parameter names.. isn't that another +1 for having a separate annotation for default values?

pontusmelke commented:
@boggle There is no such thing as  _java 8 reflection to infer parameter names_, parameter names are only available if you compile with debug flags (`-parameters` I think it is and that is most probably not supported across jvms), that is the reason why we added `@Name` in the first place. Can we focus on this PR now and add feature requests somewhere else :)

pontusmelke committed:
Fix failing windows build

boggle commented on community/kernel/src/main/java/org/neo4j/procedure/UserFunction.java:
```diff
@@ -0,0 +1,139 @@
+ *      WRITE   allows reading and writing the graph
+ *      SCHEMA  allows reading the graphs and performing schema operations
+ *      DBMS    allows managing the database (i.e. change password)
+ *
```
I thought we had said that functions are read only? Is the documentation out of sync?

boggle commented on enterprise/backup/data/databases/to/logs/debug.log:
```diff
@@ -0,0 +1,602 @@
+2016-07-04 06:17:40.643+0000 INFO  [o.n.k.i.f.CommunityEditionModule] No locking implementation specified, defaulting to 'forseti'
```
Please remove this file

boggle commented on enterprise/core-edge/src/main/java/org/neo4j/coreedge/discovery/procedures/AcquireEndpointsProcedure.java:
```diff
@@ -36,8 +36,9 @@
 import org.neo4j.helpers.collection.Iterators;
 import org.neo4j.kernel.api.exceptions.ProcedureException;
 import org.neo4j.kernel.api.proc.CallableProcedure;
+import org.neo4j.kernel.api.proc.Context;
```
Please also remove all the binary enterprise/backup/*\* files committed above (can't comment there so have to write this here)

boggle commented on integrationtests/src/test/java/org/neo4j/procedure/UserFunctionIT.java:
```diff
@@ -0,0 +1,914 @@
+        }
+
+        @UserFunction
+        public Node readOnlyCallingWriteFunction()
```
We should not support functions that perform any data writes or schema updates. Functions must run with a graph database facade that does not allow performing updates.

pontusmelke replied:
This is here to test just that, i.e. that it is not possible to perform updates in a function.

pontusmelke replied:
Oh I see what you mean the test is somewhat confusing but not completely meaningless. It asserts that we cannot get updates through a backdoor by calling another function that performs the actual update. There is another test that checks so that we cannot perform updates by calling a procedure that performs the actual update.

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/executionplan/ProcedureCallMode.scala:
```diff
@@ -34,53 +34,67 @@ object ProcedureCallMode {
     builder.result().iterator
   }
+
+    override def callFunction(ctx: QueryContext, name: QualifiedName, args: Seq[Any]) =
```
This must not exist. Functions are not allowed to update the graph.

boggle commented:
@pontusmelke I did a last round of review and left some comments. The main remaining issue is that it still looks like we have support for writing/schema updating functions in there. Didn't you say that you had it restricted to read-only? I think even for internal functions, it does not make sense to support graph updates as that breaks Cypher expression semantics. Would love to get this in next week but we need to disable updates to do so. Perhaps I also missed where we fail if updates are performed, please point me to the right spot in that case.

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/DelegatingQueryContext.scala:
```diff
@@ -181,18 +181,30 @@ class DelegatingQueryContext(val inner: QueryContext) extends QueryContext {
     singleDbHit(inner.callReadOnlyProcedure(name, args, allowed))
 
-  override def callReadWriteProcedure(name: QualifiedProcedureName, args: Seq[Any], allowed: Array[String]) =
+  override def callReadWriteProcedure(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/DelegatingQueryContext.scala:
```diff
@@ -181,18 +181,30 @@ class DelegatingQueryContext(val inner: QueryContext) extends QueryContext {
+  override def callReadOnlyFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
+    singleDbHit(inner.callReadOnlyFunction(name, args, allowed))
+
+  override def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/DelegatingQueryContext.scala:
```diff
@@ -181,18 +181,30 @@ class DelegatingQueryContext(val inner: QueryContext) extends QueryContext {
+  override def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
+    singleDbHit(inner.callReadWriteFunction(name, args, allowed))
+
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/QueryContext.scala:
```diff
@@ -156,13 +156,20 @@ trait QueryContext extends TokenContext {
+
+  def callReadOnlyFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
+
+  def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/QueryContext.scala:
```diff
@@ -156,13 +156,20 @@ trait QueryContext extends TokenContext {
+  def callReadOnlyFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
+
+  def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
+  def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/QueryContext.scala:
```diff
@@ -156,13 +156,20 @@ trait QueryContext extends TokenContext {
+  def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
+  def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
+
+  def callDbmsFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/test/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/QueryContextAdaptation.scala:
```diff
@@ -122,13 +122,22 @@ trait QueryContextAdaptation {
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any],
+                                       allowed: Array[String]): AnyRef = ???
+
+  override def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef = ???
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/test/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/QueryContextAdaptation.scala:
```diff
@@ -122,13 +122,22 @@ trait QueryContextAdaptation {
+
+  override def callDbmsFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef = ???
+
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any],
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/test/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/QueryContextAdaptation.scala:
```diff
@@ -122,13 +122,22 @@ trait QueryContextAdaptation {
+
+  override def callReadOnlyFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef = ???
+
+  override def callDbmsFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]): AnyRef = ???
```
Please remove

boggle commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/ExceptionTranslatingQueryContextFor3_1.scala:
```diff
@@ -129,18 +129,30 @@ class ExceptionTranslatingQueryContextFor3_1(val inner: QueryContext) extends Qu
+  override def callReadOnlyFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
+    translateException(inner.callReadOnlyFunction(name, args, allowed))
+
+  override def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

boggle commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/ExceptionTranslatingQueryContextFor3_1.scala:
```diff
@@ -129,18 +129,30 @@ class ExceptionTranslatingQueryContextFor3_1(val inner: QueryContext) extends Qu
+  override def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
+    translateException(inner.callReadWriteFunction(name, args, allowed))
+
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

boggle commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/compatibility/ExceptionTranslatingQueryContextFor3_1.scala:
```diff
@@ -129,18 +129,30 @@ class ExceptionTranslatingQueryContextFor3_1(val inner: QueryContext) extends Qu
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
+    translateException(inner.callSchemaWriteFunction(name, args, allowed))
+
+  override def callDbmsFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

boggle commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -636,6 +636,46 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+    callFunction(name, args, transactionalContext.statement.readOperations().functionCallRead, revertable.foreach(_.close))
+  }
+
+  override def callReadWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) = {
```
Please remove

boggle commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -636,6 +636,46 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+                  revertable.foreach(_.close))
+  }
+
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) = {
```
Please remove

boggle commented on community/cypher/cypher/src/main/scala/org/neo4j/cypher/internal/spi/v3_1/TransactionBoundQueryContext.scala:
```diff
@@ -636,6 +636,46 @@ final class TransactionBoundQueryContext(val transactionalContext: Transactional
+    callFunction(name, args, transactionalContext.statement.schemaWriteOperations().functionCallSchema, revertable.foreach(_.close))
+  }
+
+  override def callDbmsFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) = {
```
Please remove

boggle commented on community/cypher/cypher-compiler-3.1/src/main/scala/org/neo4j/cypher/internal/compiler/v3_1/spi/DelegatingQueryContext.scala:
```diff
@@ -181,18 +181,30 @@ class DelegatingQueryContext(val inner: QueryContext) extends QueryContext {
+  override def callSchemaWriteFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
+    singleDbHit(inner.callSchemaWriteFunction(name, args, allowed))
+
+  override def callDbmsFunction(name: QualifiedName, args: Seq[Any], allowed: Array[String]) =
```
Please remove

pontusmelke commented:
@boggle nicely spotted. Those methods are no longer used, just forgot to remove them.

pontusmelke committed:
Remove unused call functions

The SPI still exposed methods for calling write and schema functions which are now removed.

pontusmelke commented:
@boggle I removed all remnants (I hope) of updating functions and removed the unwanted files.

On failing functions that tries to perform updates:
- There are multiple tests for this, [here](https://github.com/pontusmelke/neo4j/blob/c096c1f265042debc3d545c444117014bcf60467/integrationtests/src/test/java/org/neo4j/procedure/UserFunctionIT.java#L305) for example.
- It works the same way it works for procedures, we put restrictions on the transaction in [OperationsFacade](https://github.com/pontusmelke/neo4j/blob/901d8e8f36c2ebc7bbf6e5f624afebe0c3161dd0/community/kernel/src/main/java/org/neo4j/kernel/impl/api/OperationsFacade.java#L1517).

pontusmelke committed:
Removed files commited by mistake

frant-hartm commented:
Hi,

I had a chance to try this out and @jexp suggested to post feedback here.

First of all - thank you for this - this finally provides real benefit in terms of readability (and capability) of Cypher. When there was first mention of "user procedures" I thought this is it.

Some notes / questions:
- how do I define an aggregation function?
- creates new instance for each invocation - not ideal when one wants to call a function for each node or relationship in a traversal.. (but procedures might want to keep state, so something akin to singleton / prototype would be great)
- would like to see constructor injection for @Context fields

And one bug/missing feature - I tried to test this using neo4j-harness, looks like it hasn't been updated yet:

`public Neo4jRule neo4j = new Neo4jRule()
            .withProcedure( MyFunction.class );`
doesn't setup the function properly
I get
`org.neo4j.driver.v1.exceptions.ClientException: Unknown function 'my.product' (line 1, column 37 (offset: 36))`

The rule works fine with a procedure in the same class and the function works when I deploy it to neo4j, so I expect there is something missing in neo4j-harness.

petraselmer commented:
changelog: Added user-defined functions
